optimum/habana/transformers/generation/utils.py

```diff
@@ -625,18 +625,28 @@ def generate(
                 # only pad if bucket_size < -1. If we are bucketing (bucket_size > 0), then that is taken care in greedy_search()
                 if not is_greedy_or_beam_and_bucket:
                     # token_idx is the current index in the generation process, it is incremented each time a new token is generated
-                    token_idx = inputs_tensor.shape[-1]
+                    token_idx = inputs_tensor.shape[1]
                     model_kwargs["token_idx"] = torch.tensor(token_idx, device=inputs_tensor.device)
                     model_kwargs["token_idx_cpu"] = token_idx
-                    if generation_config.max_new_tokens is None:
-                        generation_config.max_new_tokens = generation_config.max_length - token_idx
```

Comment:
Setting `max_new_tokens` and `max_length` in generation config is discouraged by transformers